migration-engine/ARCHITECTURE.md

```diff
@@ -0,0 +1,265 @@
+  have downtime in the first place.
+- You have a small enough data set that the rollback will not take hours/bring
+  down your application by locking tables. On top of the stress already imposed
+  by the bad deployment.
+- The down migration actually works. Are your down migrations tested?
+
+In short, down migrations give you a sense of security, but it is a false sense
+of security.
+
+- In production, currently, we will diagnose the problem for you, but rollbacks
+  are manual: you use `migrate resolve` to mark the migration as rolled back or
+  forward, but the action of rolling back is manual. So it _is_ supported, not
+  just as convenient and automated as the rest of the workflows.  Down
+  migrations are somewhat rare in real production scenarios, and we are looking
+  into better ways to help users recover from failed migrations.
+
+  There are two major avenues for migration tools to be more helpful when a
+  deployment fails:
+
+  - Give a short path to recovery that you can take without messing things up
+    even more in a panic scenario
+
+  - Guide you towards patterns that can make deploying and recovering from bad
+    deployment painless, i.e. forward-only thinking, expand-and-contract
+    pattern, etc.
+
+We're looking into how we can best help in these areas. It could very well mean
+we'll have down migrations (we're hearing users who want them, and we
+definitely want these concerns addressed).
+
+What we recommend instead of relying on down migrations:
+
+- Expand and contract 
+- Roll-forward in case of failure
+
+Rolling back schema migrations because the corresponding application code has
+to be rolled back is an anti-pattern.
+
+The view that guided Migrate development is that _some level_ of awareness of
+the migration process (schema and data) will always be required from developers
+on large enough projects. What we can do is build tools that help people
+getting their migrations right, but it can't be completely automated. There's a
+lot more to do, now that we have a stable, production-ready foundation.
+
+### What features in Migrate rely on the shadow database?
+
+The only core feature of Migrate that relies on the shadow database is
+**generating migrations**. See `How does Migrate use the shadow database` below
+for more details.
+
+### How does Migrate use the shadow database?
+
+The shadow database is the only mechanism by which Migrate can determine what
+migrations do. From Migrate's perspective, Migrations are black boxes: we do
+not parse SQL, and we want to support all database features that cannot be
+represented in the PSL yet. The only way to figure out what the effect of a
+migration is is to run it. This is necessary when we need to figure out the
+current state of a migrations history.
+
+Here is how we generate a new migration:
+
+- Create an empty shadow database, or try to make the one provided with the
+  `shadowDatabaseUrl` datasource param empty.
```

Comment:
```suggestion
- If `shadowDatabaseUrl` datasource param is set in Prisma schema, use that database and empty its content. By default create an empty shadow database with a random name like `prisma_migrate_shadow_db_c8d627ed-600d-404c-aa7c-02ecd435ffa0`, 
```

Reply:
Sorry, I needed to merge fast to post there: https://prisma-company.slack.com/archives/CGMCB82N8/p1636039544063000?thread_ts=1636039400.062700&cid=CGMCB82N8

I'm going to expand the document tomorrow, I'll take up the suggested change there :+1: